Add no_data_dependent_graph_break mode

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147342

This adds a strict mode `TORCHDYNAMO_UNBACKED_STRICT` to prevent graph breaking when we guard on data dependent. This is a better UX for those who are actively trying to make their model more dynamic, but aren't close enough to full graph to use that flag directly.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
Add unbacked strict mode that errors instead of graph breaking on data dependent errors

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Add unbacked strict mode that errors instead of graph breaking on data dependent errors"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka commented on torch/fx/experimental/_config.py:
```diff
@@ -3,6 +3,8 @@
 
 
+# [@compile_ignored: debug] Fails hard instead of graph breaking on guard on data dependent errors.
+unbacked_strict = os.environ.get("TORCHDYNAMO_UNBACKED_STRICT", "0") == "1"
```
1. mm shall the default be 1?
2. i think if we can get away without a global flag and just the default is not graph breaking is even better. 
3. I thought strict mode is the mode where fail at specialization?
TORCHDYNAMO_UNBACKED_STRICT -> TORCHDYNAMO_UNBACKED_GRAPH_BREAK?
i was going to propose mark_unbacked(..., graph_break=False)
but i guess if we end up not able to do (2) above we would need this global flag for non user triggered unbacked?

bobrenjc93 replied:
As discussed in chat, global breaks BC.

In general I consider strict a mode where we are more strict, not dissimilar to hack or JavaScript's 'use strict'.

laithsakka replied:
I see but since making setting this flag to 1 TORCHDYNAMO_UNBACKED_STRICT=1 does not actually force all unbacked to be strict but only effect the graph break Behaviour. (like it does not make an unbacked variable throw at specialization).
I strongly feel we should not use this name

bobrenjc93 replied:
That's a fair point. Ok I renamed it to `no_data_dependent_graph_break`

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5902,7 +5910,7 @@ def _make_data_dependent_error(
         *,
         size_oblivious_result: Optional[sympy.Basic] = None,
-    ) -> GuardOnDataDependentSymNode:
+    ) -> Union[GuardOnDataDependentSymNode, GuardOnDataDependentSymNodeStrict]:
```
I wonder if it make sense to just avoid changugn the error to user error here
```
 def evaluate_expr(self, output_graph=None):
        try:
            return guard_scalar(self.sym_num)
        except GuardOnDataDependentSymNode as e:
            raise UserError(  # noqa: B904
                UserErrorType.ANTI_PATTERN,
                f"Consider annotating your code using torch._check*(). {str(e)}",
                case_name="constrain_as_size_example",
            )

```
instead of introducing a new error type

laithsakka reviewed (commented):
left some comments, also prob a unit test is easy to add 
```

def fn(x, y):
    if(x.size()[0]==y.size()[0]):
        return x*100
    else:
        return x*600

@torch.compile(fullgraph=True)
def fun2(x, y):
    z =fn(x, y)
    return z, x*10, y*10

x = torch.ones(5, 5)
torch._dynamo.decorators.mark_unbacked(x, 0)
torch._dynamo.decorators.mark_unbacked(x, 1)
y = torch.randn(5, 5)
print(fun2(x, y))

```

bobrenjc93 committed:
Update on "Add unbacked strict mode that errors instead of graph breaking on data dependent errors"


This adds a strict mode `TORCHDYNAMO_UNBACKED_STRICT` to prevent graph breaking when we guard on data dependent. This is a better UX for those who are actively trying to make their model more dynamic, but aren't close enough to full graph to use that flag directly.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Add unbacked strict mode that errors instead of graph breaking on data dependent errors"


This adds a strict mode `TORCHDYNAMO_UNBACKED_STRICT` to prevent graph breaking when we guard on data dependent. This is a better UX for those who are actively trying to make their model more dynamic, but aren't close enough to full graph to use that flag directly.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka reviewed (approved):
looks good thanks ! can you update title before landing to reflect the new mode name

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 committed:
Update on "Add no_data_dependent_graph_break mode"


This adds a strict mode `TORCHDYNAMO_UNBACKED_STRICT` to prevent graph breaking when we guard on data dependent. This is a better UX for those who are actively trying to make their model more dynamic, but aren't close enough to full graph to use that flag directly.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 commented:
@pytorchbot merge -i
